dash-pipeline/bmv2/dash_acl.p4

```diff
@@ -44,6 +51,42 @@ match_kind {
         default_action = deny; \
         counters = ## table_name ##_counter; \
     }
+#endif // ARCH_BMV2_V1MODEL
+#ifdef ARCH_DPDK_PNA
+#ifdef DPDK_SUPPORTS_DIRECT_COUNTER_ON_WILDCARD_KEY_TABLE
+#error "See comments in dash_acl.p4 for code changes to make"
+#endif
+// After this issue has been fixed:
+// https://github.com/p4lang/p4c/issues/3868
+
+// (1) Add the following line as the first line of the definition of
+// ACL_STAGE:
+//    DirectCounter<bit<64>>(PNA_CounterType_t.PACKETS_AND_BYTES) ## table_name ##_counter; \
+
+// (2) Add the following line immediately after the line defining the
+// default_action of the table:
+//        pna_direct_counter = ## table_name ##_counter; \
+
+#define ACL_STAGE(table_name) \
+    @name(str(table_name##:dash_acl_rule|dash_acl)) \
+    table table_name { \
+        key = { \
+            meta. ## table_name ##_dash_acl_group_id : exact @name("meta.dash_acl_group_id:dash_acl_group_id"); \
+            meta.dst_ip_addr : LIST_MATCH @name("meta.dst_ip_addr:dip"); \
+            meta.src_ip_addr : LIST_MATCH @name("meta.src_ip_addr:sip"); \
+            meta.ip_protocol : LIST_MATCH @name("meta.ip_protocol:protocol"); \
```

Comment:
Shouldn't "protocol" be exact match?  Why is this mapping to ternery?

Reply:
The original code used match_kind optional (either complete exact match, or wildcard on all bits).  Should it have been exact match?  I am not claiming what I am writing is the final version.  Again, perhaps the original author can answer why it was the way it was in the code before this PR?

Reply:
This could have been a cut-and-paste issue :-)  I am a co-author of some of this code.  In the 5-tuple, the IP-Proto (protocol field) should have been an exact match.

Reply:
Do ALL DASH rules always have an exact match on IP protocol?  That is, is it disallowed to attempt to create a rule that matches on IP SA and DA only, and wildcard everything else?

Reply:
This is a great point.  Perhaps, in order to cover the wildcard case, this field was marked ternery.

Reply:
If you search for the first occurrence of the string "ACL_LEVEL1" on this page https://github.com/sonic-net/DASH/blob/main/documentation/general/sdn-pipeline-basic-elements.md you can see a couple of example configurations where the protocol contains `*`, which appears to be a wildcard match on the packet protocol field.  If those examples are intended to be supported by DASH, then it appears the data plane needs a way for protocol to be completely wildcarded, or completely exact match.  In P4, ternary is more general than that.  There is another match kind called optional that is exactly that.  It is not clear to me from that example whether the control plane would ever take advantage of the ability to configure more than one exact match protocol id for a single rule.